### ISSUE TITLE
When sharing a view for a contact group : Error message "No layout found for view_id :"

### ISSUE THREAD

**Issue by Jordan91** [centreon/centreon#4313]
When sharing a view for a contact group : Error message "No layout found for view_id :"

![widget 2](https://cloud.githubusercontent.com/assets/16045498/14323863/9fbd4502-fc22-11e5-96f4-84cd8ea7e937.png)

**Comment by querwin** [comment]
Hello,

Which version of Centreon are you using ?

This bug has been solved in Centreon 2.7.3

Regards,

**Comment by Jordan91** [comment]
Hi Querwin, 

I'm in 2.7.3 , widgets are up to date. 
When the user in the contact group add view , he sees this message.

**Comment by querwin** [comment]
I've just tested and you're right.

This has been fixed for contact but not for contact group maybe.

**Comment by lpinsivy** [comment]
The system of sharing view with contactgroup is not well implemented

Interface must share direclty with all contacts linked to group

**Comment by kadrika** [comment]
I just tested with our own monitoring, and we have the same problem: No layout found for view_id : 3

And we are using Centreon 2.7.3

When can we expect this to be fixed?

**Comment by 12move** [comment]
Hello everybody,

I am using Centreon 2.7.4 & the bug does not seem fixed at this stage : No layout found for view_id
It is not possible to share views to a whole groupe, what a pitty :(

Can we expect a super debug soon ? :)

Thanks!

**Comment by 12move** [comment]
Hello Guys,

I decided to take a look at the source code and to patch the issue myself.
Globally, I fixed the problem (according to me). It fits now to my uses of the Centreon tool.
Maybe someone of the Centreon commmunity can check my modifications.

About the fix...
In the php class/usr/share/centreon/www/class/centreonCustomView.class.php, you will find the generated error / function getLayout($viewId)

> throw new CentreonCustomViewException(sprintf('No layout found for view_id : %s', $viewId));

The problem is that the multidimensional array $views[$viewId] does not contain the key 'layout' ($views[$viewId]['layout']) cause the SQL statement returned an empty result (when calling the internal function getCustomViews()).

Here is my modification (function getCustomViews()):
```
 public function getCustomViews()
    {
        if (!isset($this->customViews)) {
            $query = "SELECT cv.custom_view_id, name, layout, is_owner, locked, user_id, usergroup_id, public
                          FROM custom_views cv, custom_view_user_relation cvur
                          WHERE cv.custom_view_id = cvur.custom_view_id
                          AND (cvur.user_id = " . $this->db->escape($this->userId);
            if (count($this->userGroups)) {
                $cglist = implode(",", $this->userGroups);
                $query .= " OR cvur.usergroup_id IN ($cglist) ";
            }
            
            /*You have to delete the AND is_consumed = 1*/
            //$query .= ") AND is_consumed = 1 ORDER BY user_id, name";
            $query .= ") ORDER BY user_id, name";
            $this->customViews = array();

```

This modification allows to fix two problems:

- The problem of sharing a view to a whole contact group
- It allows now to deploy automatically the shared views on the Centreon client account (without having to add the views manually). It is possible in both cases : when a view is shared for a group or a user.

If I understand correctly, the  `is_consumed ` attribute allows to specify if the view where added or not in the user account.
By removing this, this allows the automatic deployment of views on the specified shared groups or users.

Hope it will help.

Best regards,
Adrien H.

**Comment by Alex33150** [comment]
Hi guy
Your solution is functional and the bug is fixed with to this change.
Thank you for your work .

**Comment by AKaefer** [comment]
I have 2.7.6 and still have this error...

**Comment by itsul** [comment]
Please test this adjustment in detail. We had some trouble after it.

e.g.
shared widgets could be deleted globaly from every user if the option "locked" is set to no. If you set the option to "yes" the user can not delete the widget from his own view.

Result: We switched back to the orignal file.

**Comment by bitbull06** [comment]
https://github.com/centreon/centreon/issues/4313#issuecomment-234608199 seems to fix the problem in version 2.7.7
NB: all views are set to 'locked' here, so we don't mind the 'read' user getting the views per default

**Comment by lpinsivy** [comment]
This solution does'nt allows to delete view.

If I apply this patch, I can add shared view but I can't no more delete it.

**Comment by Madivil** [comment]
Good morning,

Any news on this bug?

Thanks for your answer.

Best regards,

**Comment by Dacchi** [comment]
Hello,

Several bug fixes (including this one) and improvements have been made on the custom views and will be available on the next version of Centreon (2.8.5).

This version will shortly be available.

**Comment by Madivil** [comment]
Thank you very much for the answer.

Best Regars,
